Add bootstrap5 theme

EreMaijala opened the pull request:
This is initial work for adding a bootstrap5 theme that uses Bootstrap 5.x. and to include enough back-compatibility support to make it possible to inherit other themes from it. Still pretty much work in progress but comments welcome.

Principles:
- Keep template changes from bootstrap3 to minimum, and only do string replaces
- Add bs3-compat.scss and bs3-compat.js as a compatibility layer

Here are the most significant template changes:

data-toggle="collapse" => data-bs-toggle="collapse"
data-toggle="dropdown" => data-bs-toggle="dropdown"
data-target => data-bs-target
data-dismiss="modal" => data-bs-dismiss="modal"

Other things worth noting:
- `visually-hidden` replaces the `sr-only` class, but the latter is still available via bs3-compat.scss
- font-awesome has been updated (`$fa-font-path` has changed too).
- popover, tooltip, collapse and modal are different, so the following files have differences between bootstrap3 and bootstrap5: 
  - account_ajax.js
  - cart.js
  - channels.js
  - facets.js
  - lightbox.js
- Where bootstrap3 uses the `in` class, bootstrap5 uses `open`
- legend does not have huge bottom margin in bootstrap5
- dark theme does not work properly at the moment
- header is a bit different and needs manual adjustment from bootstrap3
- Tabs now have the `active` class on the `a` element instead of `li` (bs3-compat.js handles some of this, but style changes may be required as well).
- `npm run build:scssonly` can be used to compile scss for scss-only themes (bootstrap5, sandal5). Also `npm run build:css` does that.
- compiling scss requires `npm install` to be run in the theme directory first

TODO:
- [x] Commit templates before merging.
- [x] Do we want to clean up LESS/SCSS conditionals from scss files or keep them for parity with bootstrap3/sandal?
- [x] Ensure that there are no conflicts with Bootstrap 5 and VuFind themes regarding accessibility (Wave and/or Axe could be useful tools)
- [x] [Document](https://vufind.org/wiki/development:architecture:user_interface:bootstrap5#template_changes) changes required to templates
- [x] [Document](https://vufind.org/wiki/development:architecture:user_interface:bootstrap5#sass_only) changes required with less to sass transition
- [x] Update channels.js with latest changes from bootstrap3
- [x] Enable check-bootstrap5 in QA before merging.
- [x] When merging, add changelog note about theme transition plan (bootstrap5 = beta in 10.0; bootstrap3 will be deprecated in 10.1 and bootstrap5/sandal5 promoted to default; bootstrap3 and its children will be removed in 11.0)

TODO (Issues):
- [x] [White on white text](https://github.com/vufind-org/vufind/pull/3222#issuecomment-2064030558) - VuFind link at top left and theme/language dropdowns appear as white-on-white (Bootstrap5 theme only).
  - Should be fixed.
- [x] [Similar Items](https://github.com/vufind-org/vufind/pull/3222#issuecomment-2064120165) - formatted incorrectly in sidebar and record tabs (Bootstrap5 and Sandal5).
  - Should be fixed.
- [x] [Field selection dropdown in search box](https://github.com/vufind-org/vufind/pull/3222#issuecomment-2064410537) - out of alignment (Bootstrap5 and Sandal5).
  - Should be fixed.
- [x] [Print Record](https://github.com/vufind-org/vufind/pull/3222#issuecomment-2064504199) - formatting is strange (Bootstrap5 and Sandal5).
  - Should be fixed.
  - I'm still seeing the issue in both Chrome and FF, with cleared caches - ST 
  - Oops, I was looking at the actual print results, fixed again. -EM
  - Big improvement! Remaining nitpick: I've posted an updated screenshot below showing that **text color** still varies in print preview (everything is monochrome in legacy themes) so the last thing to possibly fix is text color. - ST
  - Should be fixed as well now (BS5 makes color utils !important by default, had to disable that). -EM
- [x] [Format and availability lozenges](https://github.com/vufind-org/vufind/pull/3222#issuecomment-2064735378) - font too small (Sandal5, when compared with Sandal).
  - Should be fixed. -EM
- [x] [Floating "back"](https://github.com/vufind-org/vufind/pull/3222#issuecomment-2064758161) - extra link text at top of filter columns (Bootstrap5 and Sandal5).
  - Should be fixed (was a problem in dev branch, see #3573).
  - I'm still seeing the issue in both Chrome and FF, with cleared caches - ST 
  - Missed a file in commit, should be properly fixed now. -EM
- [x] [Main VuFind logo](https://github.com/vufind-org/vufind/pull/3222#issuecomment-2066651170) - The main VuFind logo needs more emphasis and possibly the registered trademark symbol (Bootstrap5 needs bold/etc., both Bootstraps lack the registered trademark symbol) 
  - Underlining and a bit more emphasis added. -EM
- [x] [Print preview for multi-selected items in Favorites list](https://github.com/vufind-org/vufind/pull/3222#issuecomment-2066697629) - Print view for selected items in Favorites lists contains graphics that don't render correctly. (Bootstrap5 and Sandal5)
  - Should be fixed now along with the other print issues. -EM
- [x] [Record tabs ghost highlighting](https://github.com/vufind-org/vufind/pull/3222#issuecomment-2066720008) - Previously selected tab still looks selected (but with slightly different styling than the actually selected tab). (Bootstrap5 and Sandal5)
  - Should be fixed. -EM
- [x] [Contrast in form element borders](https://github.com/vufind-org/vufind/pull/3222#issuecomment-2067092225) - Borders outside search boxes and other form elements are not contrasty enough. (Sandal5, and also Bootstrap5)
  - Focus outline for sandal5 fixed, but I can't see the other issues. -EM
  - I see a small difference in the two bootstrap versions, but I think the bootstrap5 version is slightly MORE contrasty. No change needed (unless someone who knows accessibility best practices better says differently.)
- [x] [Hover behavior in header](https://github.com/vufind-org/vufind/pull/3222#issuecomment-2067120306) - Links in header need updates to styling in new themes (Bootstrap5 especially; also Sandal5)
  - Should be much better now. However, the hover highlight is not quite the same between bootstrap3 and bootstrap5, and I'm not sure it should be either. Now everything has the same hover effect. -EM
- [x] [Advanced search limit options](https://github.com/vufind-org/vufind/pull/3222#issuecomment-2067141364) - Height of this element needs to be increased. (Bootstrap5 and Sandal5)
  - Fixed. -EM
  - Yes, you fixed those three... but I just noticed that in [New Items](http://localhost/vufind_test/Search/NewItem?ui=sandal5), there's a similar element for Department that needs to be expanded in the same way in the new themes.
  - Now fixed as well. -EM
- [x] [Links in Channels don't work](https://github.com/vufind-org/vufind/pull/3222#issuecomment-2067249367) - Links in the pop-out item menus don't work when clicked. (Bootstrap5 and Sandal5)
  - Fixed. -EM
- [x] [Problematic click behavior in Your Account menus](https://github.com/vufind-org/vufind/pull/3222#issuecomment-2070030903) - Clicking a menu item results in illegible black boxes (Bootstrap5, in contrast with Bootstrap3)
  - Fixed. -EM
- [x] [Mouseover is black in Browse and Alphabrowse](https://github.com/vufind-org/vufind/pull/3222#issuecomment-2070314783) - Hovered-over menu items are illegible due to black highlighting; in legacy version, highlighting is gray. (Bootstrap5)
  - Fixed. I agree that the highlight color is odd, but it's equally odd in old sandal. -EM
  - I'm still seeing hovered-over menu items as having black highlighting in bootstrap5 - even on a browser I've never used before. Can you check?
  - Had to tweak nav-link colors and make the navbar ones a special case instead of the other way round. -EM
- [x] [Narrow dropdown on System Maintenance page in Admin area](https://github.com/vufind-org/vufind/pull/3222#issuecomment-2077629439) - The dropdown box in the last row on this page is not wide enough to display the word "standard" in any of the themes.
  - Fixed. The width is now relative to the font size, and padding has been reduced to allow for more text in select fields in general. -EM 
  - On recheck, I see that the entire word "standard" is now visible in bootprint3 and in sandal5. However, the final 'd' is cut in half in the two bootstraps, and in sandal "standard" ends in the middle of the 'r.' So, a little more tweaking? -ST
  - Tweaked more for bootstrap5 and sandal5. I'm avoiding touching the old themes here. -EM
  - The word "standard" is fully visible now in bootstrap5 and sandal5! This is done. :)
- [x] [Extra X in search box while typing](https://github.com/vufind-org/vufind/pull/3222#issuecomment-2121094458) - There's an extra X to clear your search in the searchbox. (Bootstrap5 and Sandal5)
  - Should be fixed now. -EM
  - Confirmed - fixed! - ST
- [x] [Facets filtering lightbox layout](https://github.com/vufind-org/vufind/pull/3222#issuecomment-2121112211) - This lightbox has issues in the new themes - see below. (Bootstrap5 and Sandal5)
  - Should be fixed now. -EM
  - Confirmed - fixed! - ST
- [x] [Responsive table for search history](https://github.com/vufind-org/vufind/pull/3222#issuecomment-2121150448) - The enhancements from [PR 3204](https://github.com/vufind-org/vufind/pull/3204) are not fully implemented in the new themes. (Bootstrap5 and Sandal5)
  - Should be fixed now. -EM
  - Confirmed - fixed! - ST

EreMaijala committed:
Initial commit. Very raw.

EreMaijala committed:
Fix popper install.

EreMaijala committed:
Use minified bootstrap js.

EreMaijala committed:
Navbar + modal tweaks.

EreMaijala committed:
Fix tooltips, btn-default, forms etc.

EreMaijala committed:
Fix checkboxes, paddings, control widths etc.

EreMaijala committed:
Tweaks here and there. Initial conversion of bootprint3 and sandal.

EreMaijala committed:
Add mixins and more vars to bs3-compat.

crhallberg commented:
I wonder if we can reduce the template changes even further by re-mapping the data attributes with bs3-compat.js.

```js
const dropdownElementList = document.querySelectorAll('[data-dropdown]');
dropdownElementList.forEach((dropdownToggleEl) => {
   const target = dropdownToggleEl.getAttribute("data-target");
   if (target) {
       dropdownToggleEl.setAttribute("data-bs-target", target)
   }
});

const dropdownList = [...dropdownElementList].map(
    (dropdownToggleEl) => new bootstrap.Dropdown(dropdownToggleEl)
);
```

For the templates that aren't changed from bs3, do we want to inherit those to make it clear which templates we need to update when we make changes?

EreMaijala commented:
> I wonder if we can reduce the template changes even further by re-mapping the data attributes with bs3-compat.js.

I haven't actually tried it yet, but I suspect we'd have trouble with the initialization order of the scripts. We could make bs3-compat run early, but then it could miss something. We'd also have to employ mutation observers for any dynamically loaded content. But I have a couple of other ideas for runtime patching..

> For the templates that aren't changed from bs3, do we want to inherit those to make it clear which templates we need to update when we make changes?

My aim currently is to make it so that you could run a script to update bootstrap5 from bootstrap3. This would allow the theme to be completely standalone to avoid any confusion. But inheriting bootstrap3 is certainly an option, and I'm not committed to any specific approach yet. :)

EreMaijala commented:
Regardless of the approach, we could of course leave all template files out of git if the script works. That'd make this PR quite a bit smaller.

demiankatz commented:
> Regardless of the approach, we could of course leave all template files out of git if the script works. That'd make this PR quite a bit smaller.

That would definitely make review easier, though of course we'll want to commit the files in the end. (If we're able to go with a script-based approach, we should also add a CI test similar to lessToSass to ensure nobody forgets to keep the files in sync). In any case, I definitely like the fool-proof nature of the script-based approach, if we're able to make that work... and if not, @crhallberg's idea of inheriting files that are unchanged seems like the next best thing.

sturkel89 commented:
I have run the full test suite on this PR. Here are the results:

**ERRORS!**
Tests: 2567, Assertions: 10663, Errors: 37, Failures: 8, Skipped: 64.

Please let me know if you'd like me to paste more detail here in the PR.

EreMaijala committed:
Add a phing task to create bootstrap5 templates and remove them from git.

EreMaijala committed:
Merge branch 'dev' into dev-bootstrap5

EreMaijala commented:
I've now added a `vendor/bin/phing update-bootstrap5-templates` task and removed the templates from git. Next step would be to fix pagers.

demiankatz commented:
@EreMaijala, I see that GitHub is complaining about some things in the Javascript. Are these new issues, or is it re-reporting things we had already investigated in bootstrap3?

EreMaijala committed:
Merge branch 'dev' into dev-bootstrap5

EreMaijala committed:
Tweak Mink tests to work without jQuery :focusable selector.

EreMaijala committed:
Fix cart popovers, Mink tests, pagination.

EreMaijala committed:
Fix virtual keyboard etc.

EreMaijala committed:
Add files.

crhallberg commented:
@demiankatz it looks like [Javascript that is in bootstrap3](https://github.com/vufind-org/vufind/blob/dev/themes/bootstrap3/js/embedded_record.js#L46)

EreMaijala committed:
Switch to dart-sass from deprecated node-sass.

EreMaijala committed:
Upgrade Font Awesome to get rid of deprecation warnings.

EreMaijala committed:
Add compatibility layer for old data attributes.

EreMaijala committed:
Use load path for bootstrap vendor files.

EreMaijala committed:
Fix lightbox.

EreMaijala committed:
Fix responsive utilities, add pager, tweak breadcrumbs etc.

EreMaijala committed:
Improve navbar-form.

EreMaijala commented:
With the latest updates this is now in a much better shape, though I'm sure there's still a lot I've missed so far.

demiankatz commented:
Thanks, @EreMaijala! I've challenged @sturkel89 to look around and see if she can find anything broken... :-)

sturkel89 reviewed (changes requested):
I added bootstrap5 as an option in my theme-switcher version of config.ini. 

The first thing I noticed in the test branch is the **fun new color** of the lozenge labels for format in bootstrap3! 

(Actually, I prefer the more sedate dark blue, but it's good to have a visual indicator of which theme we're viewing and which branch we're in.)
![image](https://github.com/vufind-org/vufind/assets/107194029/50656880-8d73-47a6-8f3d-189e35f0a489)

------

The second thing I noticed is that in the test branch, **the AVAILABLE lozenge seems to be missing in three out of four themes!** (On the dev branch, the AVAILABLE lozenge appears as expected in all three available themes.)

**Explanation:**
The AVAILABLE lozenge uses white text, and has background shading in bootstrap3 only. In the other three themes, it is missing its background shading. See below:

### bootstrap3:
![image](https://github.com/vufind-org/vufind/assets/107194029/28d52ffd-d52b-4a6b-862b-41b02dd519d3)


### bootprint3:
![image](https://github.com/vufind-org/vufind/assets/107194029/94cf6ed9-bd90-4d5d-ada2-e35ce12722cb)


### sandal:
![image](https://github.com/vufind-org/vufind/assets/107194029/3f6137e9-3cb4-4aa8-ba1a-473917956f64)


### bootstrap5:
![image](https://github.com/vufind-org/vufind/assets/107194029/dfa40ca9-8b70-42d1-a39f-f0f96dce5813)

@EreMaijala, maybe this is an easy fix? Thanks!

sturkel89 reviewed (changes requested):
Here's one more weirdness I noticed in bib record view, as a bonus:

### Test branch
In the test branch, bootstrap3 displays the text about tags and the **+Add Tag** "button" as two columns within the central column of the record view. The other themes display these two elements in a vertical pile.

**bootstrap3:**

![image](https://github.com/vufind-org/vufind/assets/107194029/ad313e19-8b3e-4514-90d5-1dffc287644b)

**bootprint3:**

![image](https://github.com/vufind-org/vufind/assets/107194029/cac55647-1025-48be-ad93-97dce4cc3d2a)

**sandal:**

![image](https://github.com/vufind-org/vufind/assets/107194029/00537f66-ad75-4b56-b67c-1f990f143bf4)

**bootstrap5:**

![image](https://github.com/vufind-org/vufind/assets/107194029/fd559b91-7264-48a0-9c1c-00e08848824b)

### Dev branch:
In the dev branch, every theme shows the split column:

**bootstrap3:**

![image](https://github.com/vufind-org/vufind/assets/107194029/5b5b21b1-993d-4576-ba33-37b8c1d005af)

**bootprint3:**

![image](https://github.com/vufind-org/vufind/assets/107194029/cbac6811-27e0-41e6-bec5-fefc0d8d7c58)

**sandal:** (NB: The spacing on the text in _No Tags, Be the first to tag this record!_ is wonky here. This is probably true in the test branch as well; it's only visible here because of word wrap!)

![image](https://github.com/vufind-org/vufind/assets/107194029/e75f07b9-af38-4871-bc02-9e311bd79960)

EreMaijala committed:
Fix labels and badges. Rename a couple of override files.

EreMaijala committed:
Merge branch 'dev' into dev-bootstrap5

EreMaijala committed:
Fix badge color.

EreMaijala committed:
Merge branch 'dev' into dev-bootstrap5

EreMaijala committed:
Update searchbox_controls.js from bootstrap3.

EreMaijala committed:
Use variable overrides from a11y, setup alert colors properly, add more bs3-compat variables.

EreMaijala committed:
Merge branch 'dev' into dev-bootstrap5

EreMaijala committed:
Remove bootstrap-accessibility.

EreMaijala committed:
Merge branch 'dev' into dev-bootstrap5

EreMaijala committed:
Update JS.

EreMaijala committed:
Update SCSS.

EreMaijala committed:
Merge branch 'dev' into dev-bootstrap5

EreMaijala committed:
Update JS from bootstrap3.

EreMaijala committed:
Fix account ajax tooltips.

EreMaijala committed:
Update theme config.

EreMaijala committed:
Update scss, tweak navbar and pagination.

EreMaijala committed:
Fix navbar.

Fixes responsiveness, colors and alignment.

EreMaijala committed:
Tweak update-bootstrap5-templates task.

EreMaijala committed:
Fix pager and pagination vertical margins.

EreMaijala committed:
Add list_item_selection.js.

EreMaijala committed:
Merge branch 'dev' into dev-bootstrap5

EreMaijala committed:
Update js and scss with changes in dev.

EreMaijala committed:
Fix grid breakpoints for more bs3 compatible ones.

EreMaijala committed:
Fix data-toggle update to not apply to vufind-offcanvas.

EreMaijala committed:
Add selectors to MinkTestCase.

EreMaijala committed:
Add height to .form-control in bs3-compat.

EreMaijala committed:
Change md breakpoint to align with bs3.

EreMaijala committed:
Merge branch 'dev' into dev-bootstrap5

EreMaijala committed:
Update styles and js from bootstrap3, fix search form breakpoint.

EreMaijala committed:
Fix legend, label and MyResearch menu underline.

EreMaijala committed:
Fix hierarchy icons.

EreMaijala committed:
Fix textarea height (e.g. feedback forms)

EreMaijala committed:
Merge branch 'dev' into dev-bootstrap5

EreMaijala committed:
Apply upstream changes to bootstrap5.

EreMaijala committed:
Fix btn-default background color.

EreMaijala committed:
Revert lost changes in navbar header.

EreMaijala committed:
Fix channels popovers.

EreMaijala committed:
Disable dark mode for now.

We can't support dark mode with bootstrap3 compatibility.

EreMaijala committed:
Comment out dark mode variables for now.

EreMaijala committed:
Merge branch 'dev' into dev-bootstrap5

EreMaijala committed:
Update styles from dev.

EreMaijala committed:
Fix list item selection tests.

crhallberg commented:
Gave the changes here a read today. I know that `dropdown-item` isn't required by BS5 anymore, but do we want to keep it for easier customization?

EreMaijala commented:
> Gave the changes here a read today. I know that `dropdown-item` isn't required by BS5 anymore, but do we want to keep it for easier customization?

I think it's worth keeping, but we can revisit that later.

EreMaijala committed:
Merge branch 'dev' into dev-bootstrap5

EreMaijala committed:
Add sandal5, revert sandal and bootprint3 to bootstrap3.

EreMaijala committed:
Fix navbar active item color.

EreMaijala committed:
Add build task scssonly for less-less themes and update build:css to use it.

EreMaijala committed:
Fix sandal5 compiled.css.

EreMaijala committed:
Merge branch 'dev' into dev-bootstrap5

EreMaijala committed:
Upstream changes, fix btn-default focus.

EreMaijala committed:
Add update-bootstrap3 task, rename update-bootstrap5. Tweak tasks.

Now ignores header.phtml and aborts if there are uncommitted changes in the target directory.

EreMaijala committed:
Fix QA.

EreMaijala committed:
Copy view helpers from Bootstrap3.

EreMaijala committed:
Add mentions of bootstrap5 and sandal5 to config.ini.

EreMaijala committed:
Fix Mink tests, facet storage, missing js updates.

EreMaijala committed:
Fix tags and panel titles.

EreMaijala committed:
Disable check:scss for scss-only themes.

sturkel89 commented:
In the bootstrap5 theme, the VuFind logo in the upper left corner of the screen in every type of view is present in white-on-white and therefore is invisible. The same is true for the theme's name in the theme selector dropdown and the current language in the language selector dropdown.

This quirk is not an issue in Sandal5, because that uses the **vufind(r) search.discover.share.** icon in the upper corner, and contrasting highlighting for the selected theme and language.

![image](https://github.com/vufind-org/vufind/assets/107194029/65bfd0fa-0fe6-4c96-9fdb-63cba3585543)

![image](https://github.com/vufind-org/vufind/assets/107194029/6ac03c0e-72cb-4c26-bfd7-902871bb5748)

sturkel89 commented:
**Similar Items** in the right sidebar display strangely in bootstrap5 and sandal5; the padding and separators that are present in all the other themes are missing.

**Bootstrap5**
![image](https://github.com/vufind-org/vufind/assets/107194029/223a4b87-c87d-4537-b3de-0eb9fa500186)

**Bootstrap3**
![image](https://github.com/vufind-org/vufind/assets/107194029/6b5eece3-db47-4b74-8b4f-d1928cbd0a00)

**Sandal5**
![image](https://github.com/vufind-org/vufind/assets/107194029/ec07620a-9303-4c1d-a36c-01bd36fa7737)

**Sandal**
![image](https://github.com/vufind-org/vufind/assets/107194029/1bf60bcb-d80d-4b54-8e39-c5d20b825879)

This looks even worse with more fleshed-out similar items:

![image](https://github.com/vufind-org/vufind/assets/107194029/1d7a472d-c1bd-4c1b-8608-debe5c425bef)

-----------------------

It's even weirder in the footer. The top of the section looks like this, and then there are several screens below containing row after row of single similar items with the 'no image available' icon is centered on the page.

**Bootstrap5**
![image](https://github.com/vufind-org/vufind/assets/107194029/c93f7bd0-ad4c-4eba-ad10-0a668183316b)
In **Bootstrap3**, there's a slider. I guess the slider isn't working correctly in Bootstrap5 or Sandal5.

![image](https://github.com/vufind-org/vufind/assets/107194029/82d04589-556c-4448-9ac1-9af4ecf9d99e)

sturkel89 commented:
The **field selection dropdown** in the search box at the top of results and item pages (and to a lesser extent on the home page) is a few pixels higher than it should be, in Bootstrap5 and Sandal5. Also, the text is not centered within the box.

These things are aligned perfectly in the legacy themes in the test branch.

### Examples:

**Sandal5**
![image](https://github.com/vufind-org/vufind/assets/107194029/ae0602e7-bc83-4e62-96a4-af52d1f44f8a)

**Sandal**
![image](https://github.com/vufind-org/vufind/assets/107194029/8d94c0b6-ac67-4413-849b-3def315c8e3d)

**Bootstrap5**
![image](https://github.com/vufind-org/vufind/assets/107194029/680c7c40-823d-416e-81cb-697611b47cca)

**Bootstrap3**
![image](https://github.com/vufind-org/vufind/assets/107194029/721be38b-5603-4493-bfd8-31c572bf7fb0)

sturkel89 commented:
The **Print record** view in the new themes is not working correctly. I'm comparing themes within the new branch.

**Bootstrap3 - looks normal:**
![image](https://github.com/vufind-org/vufind/assets/107194029/76ee5843-02d2-4163-abf1-46b7432b5fe0)

**Bootstrap5 - all text wants to have a white background:**
![image](https://github.com/vufind-org/vufind/assets/107194029/9cc0653a-41c5-42de-8b9c-73ba6e291a84)

I could provide examples comparing Sandal and Sandal5, but the effect is exactly the same there.

------------
### UPDATE

The shading is gone and the print preview looks much better! 

My only remaining comment is that in the legacy themes, EVERYTHING in the print preview is monochrome, while in the new themes text color can vary.  

See below - with the demo driver, there are all sorts of unusual holdings, and electronic holdings apparently present availability in green. This screenshot is from the print preview view in Sandal5. In old Sandal, all the text is black.

This probably doesn't matter, but for consistency you might change that in the code.

![image](https://github.com/vufind-org/vufind/assets/107194029/d4c1d733-d10e-4ffe-987b-5134dfbfeebd)

sturkel89 commented:
The font inside the availability and format lozenges in Sandal5 is **extremely tiny**. It's noticeably smaller than the font in the lozenges in Sandal. 

(I also prefer the green color in Sandal to the gray color in Sandal5.)

**Sandal**
<img src="https://github.com/vufind-org/vufind/assets/107194029/711a0efb-3646-4ec7-9f22-64bdab29c000" width="375">

**Sandal5**
<img src="https://github.com/vufind-org/vufind/assets/107194029/19cfaf48-edd6-4680-8c10-6df198e8c8a3" width="375">

sturkel89 commented:
### Floating "back"
There is a strange floating "back" text link at the top of the filters column in Bootstrap5 and Sandal5. This text link does not appear in the legacy themes.

**Bootstrap5**
![image](https://github.com/vufind-org/vufind/assets/107194029/b5f7b72c-4923-457c-bbcd-58a50028f92b)

**Bootstrap3**
![image](https://github.com/vufind-org/vufind/assets/107194029/37449a6f-c915-4815-9ef6-e1f7000df3a3)

**Sandal5**
![image](https://github.com/vufind-org/vufind/assets/107194029/a8d2e9fe-a182-41e4-a0cb-ba9babcc762b)

**Sandal**
![image](https://github.com/vufind-org/vufind/assets/107194029/66f54757-a70c-474d-9b51-68da3787b063)

EreMaijala committed:
Fix navbar nav links, similar items.

EreMaijala committed:
Fix list-group bottom margin.

EreMaijala committed:
Fix similar items carousel.

EreMaijala committed:
Fix search type field alignment (and other selects as well).

EreMaijala committed:
Fix print media table background.

EreMaijala committed:
Fix labels, drop a11y.scss.

EreMaijala committed:
Sort bootstrap imports.

EreMaijala committed:
Tweak badge and label font sizes (again).

EreMaijala committed:
Fix typo.

EreMaijala committed:
Merge branch 'dev' into dev-bootstrap5

sturkel89 commented:
### Main VuFind logo
The VuFind logo is now visible in the upper left corner of the screen in Bootstrap5. However, it's much less prominent than the one on Bootstrap3 which is larger, bold, and underlined to show it's a hyperlink. 

Also, do we want to add the **registered trademark symbol** in the two Bootstrap themes?

**Bootstrap5:**
![image](https://github.com/vufind-org/vufind/assets/107194029/d314e26d-324f-4091-89fc-37afedc0410a)

**Bootstrap3:**
![image](https://github.com/vufind-org/vufind/assets/107194029/35ba3615-b47b-441b-a444-b97f3436c300)

sturkel89 commented:
### Print view for multi-selections on Favorites list

When I'm in a Favorites list and select multiple items, I can ask to print that selection. The print preview display in Bootstrap5 and Sandal5 include lozenge shapes that have contrasting text on them. This was fixed in an earlier PR so that print preview has none of those graphics. Can we make that same change to this feature in these two new themes?

I think this is related to the still-present issue with Print Record; the white background surrounding text in some fields would not show if all the shading were removed.

**Bootstrap5**
![image](https://github.com/vufind-org/vufind/assets/107194029/21747385-5323-496f-90d8-70eec8eba475)

**Bootstrap3**
![image](https://github.com/vufind-org/vufind/assets/107194029/1a712e31-0f63-4dbe-b7b2-4fa3e53a88ca)

sturkel89 commented:
### Record tabs ghost highlighting

In the new themes, previous tabs remain highlighted when you switch and click on another tab. There is always one recently-viewed tab that shows highlighting and the changed text color, in addition to the currently selected tab. This occurs with both Bootstrap5 and Sandal5, and not in the legacy themes.

**Bootstrap5**
![image](https://github.com/vufind-org/vufind/assets/107194029/f7501331-5610-49c1-a6e1-15b9abe70232)

**Bootstrap3**
![image](https://github.com/vufind-org/vufind/assets/107194029/69842798-1bf0-48d9-9a96-45adc84f745c)

sturkel89 commented:
### Contrast in form element borders

The borders of search boxes and other form elements have less contrast in the new themes than in the legacy themes. 

I think the outlines of the form elements in the new themes need to be set to match those in the legacy themes on dev.

There was a PR a few months ago that led to the more contrasty versions ([PR 3207](https://github.com/vufind-org/vufind/pull/3207)) that you can reference.

**Sandal5**
<img src ="https://github.com/vufind-org/vufind/assets/107194029/682b35a5-4ddd-462f-9934-d5ffcae21cc7" width=400>

**Sandal**
<img src ="https://github.com/vufind-org/vufind/assets/107194029/8634d0c9-e688-4441-bb28-d5686bcddfcc" width=400>

**Bootstrap5**
<img src ="https://github.com/vufind-org/vufind/assets/107194029/71222875-9f91-4c54-b9fc-7a075812483f" width=400>

**Bootstrap3**
<img src ="https://github.com/vufind-org/vufind/assets/107194029/971413fe-0faf-49f7-980a-3d29f8909449" width=400>

I think the orange outline in Sandal is very visible. It's not as clear to my eye that the search box outline on Bootstrap3 is an improvement over Bootstrap5, so I'm not sure about that one. I wanted to document the difference.

-----------
### UPDATE

Thanks for fixing the issue in Sandal5! That was the important one. 

I see a DIFFERENCE between the focus outline in bootstrap3 and bootstrap5, but to my eye the bootstrap5 version is MORE visible than the other. I don't think it needs fixing, unless someone who knows more about accessibility thinks the other version is better.

sturkel89 commented:
### Hover behavior in header

With all options enabled and a logged-in user, there are four items in the header area: two links (Your Account and Log Out) and two dropdowns (Theme and Language). In the legacy themes, when you mouse over any of those four elements, you see identical contrasting shading in all cases.

In the two new themes, mouseover of the **links** behaves differently from mouseover of the **dropdowns**. 

In Bootstrap5, the shading for mouseover of the Your Account and Log Out links renders them difficult to read: white on very light gray. In Sandal, the items are legible, but it's clear that they're highlighted in a different color than the dropdowns.

**Bootstrap5**
![image](https://github.com/vufind-org/vufind/assets/107194029/e5f4a17e-d93e-4c44-be91-342f52f76097)
![image](https://github.com/vufind-org/vufind/assets/107194029/0d0d4abd-fcaf-4f69-be86-ca0ca1c257b4)


**Bootstrap3**
![image](https://github.com/vufind-org/vufind/assets/107194029/ba144c4c-f5f6-4d7f-bd68-13a599ffb567)

**Sandal5**
![image](https://github.com/vufind-org/vufind/assets/107194029/80076eb7-2653-4620-bba1-61c48f27c947)
![image](https://github.com/vufind-org/vufind/assets/107194029/08f20f32-3a00-47eb-b388-25ee687c9b21)

**Sandal**
![image](https://github.com/vufind-org/vufind/assets/107194029/a0f24830-13d5-4256-962c-c6e29624febb)

sturkel89 commented:
### Advanced Search limit selections - height needs adjustment

On the Advanced Search screen, there are three browsable "limits" available: **Call Number**, **Language**, and **Format**. 

In the legacy themes, each of these elements is set to display the **first 10 values** on the options list.

In the new themes, each of these elements is tall enough to show **one value** plus a few pixels of another value. It's still possible to scroll through the list, but I think we want these fields to be 10 rows high.

**Bootstrap5**
<img src="https://github.com/vufind-org/vufind/assets/107194029/47ddfaa7-524e-4efe-b0ba-766fcaaa6e7a" width = 500>

**Bootstrap3**
<img src="https://github.com/vufind-org/vufind/assets/107194029/f423dde5-c238-4892-bd64-acc4a25d7a77" width = 500>

**Sandal5**
<img src="https://github.com/vufind-org/vufind/assets/107194029/a8d5cb96-0559-45cb-8781-43505fcbdcbd" width = 500>

**Sandal**
<img src="https://github.com/vufind-org/vufind/assets/107194029/ff10897d-e283-4d9b-9788-090668959c26" width = 500>

-----------------------------
### EDIT: one more field to fix!

The **New Items** search page has the same three elements as Advanced Search, plus the additional field **Department** which also needs to be expanded in Sandal5 and Bootstrap5.

![image](https://github.com/vufind-org/vufind/assets/107194029/746f3d5f-e35a-497d-bcab-604f9d299dd9)

sturkel89 commented:
### Links in Channels don't work

The new themes include improved versions of the pop-out informational elements for items on Channels pages. The new version appears to the side of the item rather than below (which usually displays below the fold). It also looks more modern.

_The problem is that the links to **View Record** and **Explore related channels** can't be clicked_. (You _can_ click, but all that happens is that the pop-out menu disappears and you're still on the Channels page.

The new themes DO NOT display the bug ([VUFIND-1686](https://openlibraryfoundation.atlassian.net/browse/VUFIND-1686)) I've discovered across all of the legacy themes, which is that if you click on a given item on the Channels page, click away, and then click again, the second and subsequent viewings show an abbreviated menu with no content aside from the **View Record** and **Explore related channels** buttons. (In most cases, the links in the abbreviated menus are not clickable either.)

-----------------
Here's a comparison between the new and old full menus. 

### **Sandal:**

<img src="https://github.com/vufind-org/vufind/assets/107194029/c8c0c7d5-ce05-44ad-8dd3-8b8fb434062b" width = 400>


### **Sandal5:**

<img src="https://github.com/vufind-org/vufind/assets/107194029/4b292c7a-ce27-40ee-a1e0-9533c8b7102a" width = 400>

EreMaijala committed:
Fix offcanvas toggles, print table background.

EreMaijala committed:
Fix btn-default active and focus colors.

EreMaijala committed:
Fix active tab handling.

EreMaijala committed:
Fix sandal focus outline. Fix multi-select height.

EreMaijala committed:
Fix navbar colors, search tabs.

EreMaijala committed:
Switch to manual focus to fix buttons in popovers in channels.

sturkel89 commented:
### PROBLEMATIC CLICK BEHAVIOR IN MY ACCOUNT MENUS (bootstrap5)

### Bootstrap3
In bootstrap3, the **hover color** in the main screen dropdowns is black text on a light gray background. This differs from the appearance of the **four buttons** across the top of the screen (Your Account, Log Out, Theme, and Language), which are white text on a black button when moused over. The **active theme** is indicated by white text on a dark blue background. The **moused-over theme** or language in the dropdowns is white text on a light gray background. When you click to change theme or language, the selection retains its black on light gray appearance and then the menu disappears when the change is applied.

Screenshot of bootstrap3 appearance:
![image](https://github.com/vufind-org/vufind/assets/107194029/d7b2cb7d-02bf-4873-b537-c4563aa12848)

### Bootstrap5
The **hover color** in bootstrap5 main screen dropdowns is white text on a black background, to match the color scheme of the **four buttons** across the top of the screen (Your Account, Log Out, Theme, and Language). The **active theme** or language is white text on a dark blue background. When you click to change theme or language, the selection retains its white on black appearance and then the menu disappears when the change is applied.

In this screenshot, the active theme is bootstrap5, and the mouse just hovered over bootprint 3 so it displays in black:
![image](https://github.com/vufind-org/vufind/assets/107194029/dc02f4cb-9f28-47a4-92d6-b2a91020d3a0)

So far, so good!

Now, let's look at the **Your Account** area in the two themes.

### Bootstrap5
In bootstrap5, the **active menu item** is white text on a dark blue background, as in the top menu. The **moused over menu item** is black text on a gray background. 

![image](https://github.com/vufind-org/vufind/assets/107194029/93d6a503-5710-44f7-9917-895eec3569b6)

When you **click** on that menu item, the text becomes dark blue on a black background and is pretty much illegible.

![image](https://github.com/vufind-org/vufind/assets/107194029/8aa3b443-03a2-4425-83a8-a910417be5b8)

### Bootstrap3
In bootstrap3, the **active menu item** is white text on a dark blue background, as in the top menu. The **moused over menu item** is black text on a gray background. When you **click** on that menu item, the button retains its black text over a gray background appearance, and then the selection changes so the newly active menu item is white text on a dark blue background.

![image](https://github.com/vufind-org/vufind/assets/107194029/d64785d9-e5fa-4d40-a0b9-6fbb67e04d50)

I'd suggest that the click behavior in bootstrap5 should be changed to be the same as bootstrap3; let the button look the same upon click as it was upon mouseover, so that you don't have that moment of illegibility.

sturkel89 commented:
### Mouseover is black in Browse and Alphabrowse (bootstrap5)

Hovered over text in Browse and Alphabrowse in bootstrap5 is illegible due to being highlighted in dark text on a black box.

![image](https://github.com/vufind-org/vufind/assets/107194029/bca2f343-59b9-428e-9239-31a9300a2ab2)

It looks fine on bootstrap3:
![image](https://github.com/vufind-org/vufind/assets/107194029/86ab1be8-5d81-4a93-af43-229008339e0e)

Can you change the highlight to gray instead of black on bootstrap5? 

----------------
**Side note:** 
Hover text is fine in Sandal and sandal5, although I'm not a fan of the very bright highlight color on the active/selected items in each menu! I don't think that color appears anywhere else in VuFind themes. Feel free to ignore this editorial comment!

![image](https://github.com/vufind-org/vufind/assets/107194029/a8eea494-cd5c-45bd-826e-5696e3471bc3)

EreMaijala committed:
Fix reboot position, fix navbar dropdowns.

EreMaijala committed:
Less to sass.

EreMaijala committed:
Fix search tabs in results, disable !important color utilities to fix print.

EreMaijala committed:
Fix height of non-multiple select when size is specified.

EreMaijala committed:
Fix MyResearch menu hover color.

EreMaijala committed:
Fix normal nav-link hover color.

EreMaijala committed:
Merge remote-tracking branch 'origin/dev' into dev-bootstrap5

EreMaijala committed:
Revert sandal common.less, update cookieconsent styles.

EreMaijala committed:
Upgrade to Bootstrap 5.3.3.

EreMaijala committed:
Fix MyResearch navbar dropdown menu hover color.

EreMaijala committed:
Add sr-only-focusable to bs3-compat.

sturkel89 commented:
I looked at the Admin area in the four relevant themes. Overall, everything looks fine. There are some styling differences between the legacy themes and the new themes, but nothing radical. (I think the new themes look a bit better, with less boldface.) 

The only issue is a dropdown box in the System Maintenance window that is never wide enough to display the entire word "standard" -- it's in the bottom row. Is this something you'd fix here, so it will look better in the future? Seems way too minor to make a Jira ticket...

### Bootstrap3
![image](https://github.com/vufind-org/vufind/assets/107194029/95486993-0939-4072-9601-c8c962397943)

### Bootstrap5
![image](https://github.com/vufind-org/vufind/assets/107194029/1a2b439d-8122-4ecb-b828-5c2d8bc30b43)

### Sandal
![image](https://github.com/vufind-org/vufind/assets/107194029/901f5af6-e881-4e2d-a73b-e35f69b6e2c6)

### Sandal5
![image](https://github.com/vufind-org/vufind/assets/107194029/883be18a-7e77-4a88-b152-57155610862d)

EreMaijala committed:
Fix select field in admin module.

Reduces padding for other select fields as well.

EreMaijala committed:
Tweak admin form a bit more.

EreMaijala committed:
Merge remote-tracking branch 'origin/dev' into dev-bootstrap5

sturkel89 reviewed (approved):
All of my issues to date have been addressed - thanks, @EreMaijala ! 

I think we're in shape to merge this branch, and then we'll continue to move forward.

demiankatz commented on config/vufind/config.ini:
```diff
@@ -43,6 +43,8 @@ titleSeparator = "::"
 ;   sandal     = bootstrap3 theme with a "flat" styling applied (a newer look
 ;                than bootprint3).
+;   bootstrap5 = HTML5 theme using Bootstrap 5 with minimal styling (beta).
+;   sandal5    = Like sandal, but based on bootstrap5
```
Should we mark this beta as well?

EreMaijala replied:
Yes, I'll make sure to do that.

demiankatz reviewed (changes requested):
@EreMaijala, one nitpick below, and one larger question before we merge this:

What is the best way to keep things in sync going forward, once we merge this? Is the idea that we'll run the automatic template conversion process every time a bootstrap3 template changes? If so, should we add something to the GitHub Actions setup to test for uncommitted bootstrap5 changes so we don't forget (e.g. similar to our current lessToSass check)?

EreMaijala commented:
@sturkel89 Thanks! I'll merge the latest changes from dev at least once more so that we're fully up to date.

EreMaijala commented:
@demiankatz Something like that, or the other way round (update bootstrap3 from bootstrap5). Regardless of that, I think a check for inconsistencies would be very useful. It could also be applied to most js and scss files. At least until the themes start to diverge on purpose.

demiankatz commented:
> @demiankatz Something like that, or the other way round (update bootstrap3 from bootstrap5). Regardless of that, I think a check for inconsistencies would be very useful. It could also be applied to most js and scss files. At least until the themes start to diverge on purpose.

Great! I think at least at first, it makes sense to test from bootstrap3 --> bootstrap5, since existing work in progress is going to have bootstrap3 template changes, and it will probably be easier to make a transition if we start in that direction. At some point, we may want to decide to flip the direction once we've merged this work into all open PRs and gotten things in sync.

Let me know if you'd like any help on setting up a mechanism!

EreMaijala committed:
Apply upstream changes to bootstrap5.

EreMaijala committed:
Merge remote-tracking branch 'origin/dev' into dev-bootstrap5

EreMaijala committed:
Apply upstream changes to bs5 themes.

EreMaijala committed:
Revert bad change.

EreMaijala commented:
@demiankatz There's now phing task check-bootstrap5 that checks templates, js and scss files (apart from those known to have differences):

templates:
- header.phtml
- RecordTab/similaritemscarousel.phtml

scss:
- bootstrap.scss
- components/similar-carousel.scss
- vendor/**

js:
- account_ajax.js
- cart.js
- channels.js
- facets.js
- lightbox.js
- vendor/bootstrap-accessibility.min.js
- vendor/bootstrap.min.js

EreMaijala committed:
Fix accessibility issues with header and tabs.

EreMaijala committed:
Add templates.

EreMaijala committed:
Update .gitignore.

EreMaijala committed:
Enable check-bootstrap5 task.

EreMaijala committed:
Add templates in templates/admin/tags.

demiankatz reviewed (changes requested):
When I run the full test suite with -Dtest_theme=sandal5, I get a few errors/failures:

```
There were 2 errors:

1) VuFindTest\Mink\RecordVersionsTest::testVersions
Exception: Failed to get text after 6 attempts.

/usr/local/vufind/module/VuFind/src/VuFindTest/Integration/MinkTestCase.php:765
/usr/local/vufind/module/VuFind/src/VuFindTest/Integration/MinkTestCase.php:684
/usr/local/vufind/module/VuFind/tests/integration-tests/src/VuFindTest/Mink/RecordVersionsTest.php:68
/usr/local/vufind/module/VuFind/tests/integration-tests/src/VuFindTest/Mink/RecordVersionsTest.php:91

2) VuFindTest\Mink\RecordVersionsTest::testVersionsInSearch2
Exception: Failed to get text after 6 attempts.

/usr/local/vufind/module/VuFind/src/VuFindTest/Integration/MinkTestCase.php:765
/usr/local/vufind/module/VuFind/src/VuFindTest/Integration/MinkTestCase.php:684
/usr/local/vufind/module/VuFind/tests/integration-tests/src/VuFindTest/Mink/RecordVersionsTest.php:68
/usr/local/vufind/module/VuFind/tests/integration-tests/src/VuFindTest/Mink/RecordVersionsTest.php:101

--

There was 1 failure:

1) VuFindTest\Mink\CollectionsTest::testContextLinks
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'http://localhost/vufind_test/Collection/subcollection1/HierarchyTree#tabnav'
+'http://localhost/vufind_test/Collection/subcollection1/HierarchyTree'

/usr/local/vufind/module/VuFind/tests/integration-tests/src/VuFindTest/Mink/CollectionsTest.php:166

ERRORS!
Tests: 2816, Assertions: 17962, Errors: 2, Failures: 1.
```

Are you seeing the same thing? Is this expected?

EreMaijala committed:
Make Bootstrap3 view helpers extend Bootstrap5's to keep things in sync.

EreMaijala committed:
Merge remote-tracking branch 'origin/dev' into dev-bootstrap5

EreMaijala committed:
Apply upstream changes to bootstrap5.

sturkel89 commented:
I did some additional testing of this branch today and found a couple of new (?) issues with the bootstrap5 and sandal5 themes that don't appear in the legacy themes.

### Extra X in search box while typing

I discovered this while testing the Virtual Keyboard but then realized it's an issue even without that feature enabled. In the two new themes, an extra X appears while entering text and seeing search suggestions. There's only one X visible under the same circumstances in the three legacy themes.

**_GOOD - LEGACY THEMES_**
**Sandal**
![image](https://github.com/vufind-org/vufind/assets/107194029/b27dab9f-6289-438b-a711-053ce0eb9a6d)

**Bootstrap3**
![image](https://github.com/vufind-org/vufind/assets/107194029/e7b77ccb-947a-4886-939d-8419d68793d1)

**_NOT SO GOOD - NEW THEMES_**

**Sandal5**
![image](https://github.com/vufind-org/vufind/assets/107194029/85d8b8e0-50f3-446c-9252-742bc3bafe54)

![image](https://github.com/vufind-org/vufind/assets/107194029/6dcc84de-0f82-455c-bb51-f74360038795)

**Bootstrap5**
![image](https://github.com/vufind-org/vufind/assets/107194029/c7d5f610-2b91-487f-be21-048bd50c708b)

**Hebrew example shows different locations of the two Xes**
![image](https://github.com/vufind-org/vufind/assets/107194029/5ba349ca-06f0-4dfe-907f-15e6d1154513)

sturkel89 commented:
### Facets filtering layout

The layout of the facets lightbox looks different in the new themes vs. the legacy themes. I am not sure that this is a big problem, but it looks strange in the new themes and I think something may be wrong with the styling -- especially since the lightbox behaves better after you resize the browser.

NB: facet sorting, filtering, and clearing the search box all behave normally in every theme.

Examples come from Sandal vs. Sandal5, but the same is seen in Bootstrap3 vs. Bootstrap5.

**GOOD BEHAVIOR - LEGACY THEMES**
In sandal, the Sort and Filter boxes are aligned in a single row, and the contents of the long table are scrollable within a lightbox that is fully visible on the screen. 

When you resize the browser window in the legacy themes, the behavior of the lightbox stays the same – it just gets taller or shorter, as appropriate.

![image](https://github.com/vufind-org/vufind/assets/107194029/1af7d730-69e2-478b-b6a3-1d44e52e53ea)

**QUESTIONABLE BEHAVIOR - NEW THEMES**
In sandal5, the Sort options are on one line and the Filter box is on a second line. The contents of the table are in VERY long lightbox that doesn't fit on the page at first.

![image](https://github.com/vufind-org/vufind/assets/107194029/ad035c08-8a81-4ff3-aa4d-20d03b1415d9)

When you resize the browser window, either making it longer or shorter, the scroll bar appears on the right and the CLOSE button pops up from the bottom.

![image](https://github.com/vufind-org/vufind/assets/107194029/ee83d6ab-3c3c-4f85-b745-d13a7f214d9a)

This post-resizing appearance/behavior is more in line with what I'd expect. I would prefer that the Sort and Filter options appear on one line at the top of the lightbox in fullscreen.

sturkel89 commented:
### Responsive table in search history - not fully implemented in new themes

In reference to [PR 3204](https://github.com/vufind-org/vufind/pull/3204), each item in the search history should be displayed with information in different shades in alternating rows in mobile view. There should be space between the each "recent search", and each item should have a bit of shading to give it a slightly 3D/raised appearance

Search history looks fine in the legacy themes, but in the new themes some of this styling is missing. These screenshots show Bootstrap*, but the same is found in the two versions of Sandal.

**Bootstrap3**

![image](https://github.com/vufind-org/vufind/assets/107194029/d85a10c8-d577-47a2-8cc7-0d7b7f81acfd)

**Bootstrap5**

![image](https://github.com/vufind-org/vufind/assets/107194029/7d9c8c91-d034-4b3e-9749-2f89539aa051)

EreMaijala committed:
Fix selectors.

EreMaijala committed:
Merge remote-tracking branch 'origin/dev' into dev-bootstrap5

EreMaijala committed:
Fix Mink tests.

EreMaijala committed:
Hide browser's X button in search fields.

EreMaijala committed:
Fix facet lightbox.

EreMaijala committed:
Fix responsive tables.

EreMaijala commented:
@sturkel89 Thanks for reporting the issues. They should be fixed now.

EreMaijala committed:
Add composer task 'install-build-deps'.

demiankatz committed:
Merge branch 'dev' into dev-bootstrap5

demiankatz commented on module/VuFind/src/VuFind/View/Helper/Bootstrap3/Search.php:
```diff
@@ -38,7 +38,7 @@
  * @link     https://vufind.org/wiki/development Wiki
  */
-class Search extends \VuFind\View\Helper\AbstractSearch
+class Search extends \VuFind\View\Helper\Bootstrap5\Search
```
Did you forget to remove the code from this class? It looks to me like the contents are identical to the Bootstrap5 version, but maybe I'm missing something.

EreMaijala replied:
And I thought I even double-checked what I did... but yes, I forgot. Now fixed.

demiankatz reviewed (changes requested):
I think I may have spotted one minor oversight. Other than that, we may be ready, though I'll give @crhallberg a little longer to comment as well.

EreMaijala committed:
Remove extraneous view helper code.

crhallberg commented on build.xml:
```diff
@@ -958,4 +960,154 @@ ${git_status}
   </target>
+
+  <!-- Update bootstrap3 templates from bootstrap5 -->
+  <!-- N.B. This destroys any existing bootstrap5 templates (except explicitly excluded ones) -->
```
"N.B. This destroys any existing _bootstrap3_ templates"?

EreMaijala replied:
Fixed!

crhallberg commented on themes/bootstrap5/js/bs3-compat.js:
```diff
@@ -0,0 +1,150 @@
+/*global VuFind*/
+VuFind.register('bs3-compat', function cookie() {
```
Could we consider a more descriptive name for this function? Other ones have object-like names like ItemStatuses.

It's a shame (my shame specifically) that you needed to register a module to make sure the timing of this correct. I should add an `init` and a `ready` event to common.js.

Also, not a priority, but some of the below might be able to be moved into build.xml. Eventually.

EreMaijala replied:
Oops, that was a silly one. I changed the module and function name to a more descriptive one.

EreMaijala replied:
And yes, some of the changes could perhaps be moved to build.xml (or just done manually), but the intention here is to allow also bootstrap3-based themes with their own templates to work properly with minimal changes. In the long run I expect this to be changed to opt-in, or get deprecated and finally removed.

crhallberg commented on themes/bootstrap5/js/facets.js:
```diff
@@ -0,0 +1,301 @@
+          $facetContainer.find('.facet-load-indicator').remove();
+        });
+        VuFind.lightbox.bind($('.sidebar'));
+        VuFind.emit('VuFind.sidefacets.loaded');
```
A favor for me: this shouldn't have `VuFind.` at the front.

EreMaijala replied:
Sorry, I'm not following. This file is a direct copy from themes/bootstrap3/js/facets.js apart from the session storage handling.

crhallberg replied:
No worries, it's a mistake of mine from the listen-emit pull request.

EreMaijala replied:
Ok. :) 

This reminds me that there's a slight problem with events in the current facet lightbox code: the modal resize relies on the fact that Bootstrap 3 triggers the show.bs.modal event on each show command even when the modal is already visible, while Bootstrap 5 doesn't. I'll post a fix for that separately after this PR has been merged.

demiankatz reviewed (changes requested):
There's no new feedback from me, but I see @crhallberg has caught a few more small things, so I'm going to put this into "request changes" mode until you have a chance to address his feedback.

crhallberg commented on themes/bootstrap5/templates/header.phtml:
```diff
@@ -0,0 +1,117 @@
+            <?php if ($account->getIdentity()): ?>
+              <li class="logoutOptions<?php if ($account->dropdownEnabled()): ?> with-dropdown<?php endif ?>">
+                <a href="<?=$this->url('myresearch-home', [], ['query' => ['redirect' => 0]])?>" class="icon-link">
+                  <span id="account-icon" class="icon-link__icon" role="link"><?=$this->icon('my-account') ?></span>
```
I'm unsure about the addition of `role="link"` here. Since we're inside a link already, is it superfluous?

EreMaijala replied:
Sort of, yes, but Bootstrap's tooltip component adds, for better or worse, aria-label attribute to the span, and that's not allowed without a proper role.

crhallberg replied:
I see! Thanks for explaining!

EreMaijala committed:
Fix comment.

EreMaijala committed:
Change Bootstrap 3 compatibility JS module name.

EreMaijala committed:
Update function name too.

demiankatz reviewed (approved):
Looks like everything we expect to address here has been addressed, and there are a couple of follow-up actions we can take after merging this. Thanks, @crhallberg and @EreMaijala!

EreMaijala committed:
Merge remote-tracking branch 'origin/dev' into dev-bootstrap5
